### ISSUE TITLE
Wired results on nuisances classes

### ISSUE THREAD

**Issue by khawar-islam** [mmmmimic/OOD-CV-challenge#2]
Dear Manxi,

I have used pre-trained model like mentioned in your code. At the moment, I just pre-trained `convnext_base` for `150` epochs with batch size `64`. I am only using one model and also edited `num_classes=10
`
**Epoch 150**

```
Test shape: [ 0/16]	Time  0.768 ( 0.768)	Loss 2.2269e-01 (2.2269e-01)	Acc@1  95.31 ( 95.31)	Acc@5 100.00 (100.00)
Test shape: [10/16]	Time  0.252 ( 0.300)	Loss 5.4340e-01 (8.7980e-01)	Acc@1  89.06 ( 83.95)	Acc@5  98.44 ( 98.58)
 *   Acc@1 79.001 Acc@5 98.165
Test pose: [ 0/17]	Time  0.706 ( 0.706)	Loss 1.3559e-01 (1.3559e-01)	Acc@1  98.44 ( 98.44)	Acc@5 100.00 (100.00)
Test pose: [10/17]	Time  0.253 ( 0.294)	Loss 3.6440e-02 (6.6026e-01)	Acc@1  96.88 ( 88.92)	Acc@5 100.00 ( 98.58)
 *   Acc@1 87.535 Acc@5 98.140
Test texture: [0/9]	Time  1.047 ( 1.047)	Loss 1.3467e-01 (1.3467e-01)	Acc@1  96.88 ( 96.88)	Acc@5 100.00 (100.00)
 *   Acc@1 92.743 Acc@5 98.938
Test context: [0/9]	Time  1.063 ( 1.063)	Loss 9.8676e-02 (9.8676e-02)	Acc@1  96.88 ( 96.88)	Acc@5 100.00 (100.00)
 *   Acc@1 88.713 Acc@5 98.765
Test weather: [ 0/22]	Time  0.781 ( 0.781)	Loss 7.5742e-02 (7.5742e-02)	Acc@1  98.44 ( 98.44)	Acc@5 100.00 (100.00)
Test weather: [10/22]	Time  0.258 ( 0.302)	Loss 5.2494e-01 (2.6996e-01)	Acc@1  87.50 ( 92.33)	Acc@5 100.00 ( 99.43)
Test weather: [20/22]	Time  0.252 ( 0.278)	Loss 8.3137e-02 (4.0693e-01)	Acc@1  98.44 ( 89.58)	Acc@5 100.00 ( 99.40)
 *   Acc@1 89.614 Acc@5 99.407
Test occlusion: [ 0/91]	Time  0.767 ( 0.767)	Loss 1.2439e-01 (1.2439e-01)	Acc@1  96.88 ( 96.88)	Acc@5 100.00 (100.00)
Test occlusion: [10/91]	Time  0.253 ( 0.300)	Loss 3.4803e-01 (4.4100e-01)	Acc@1  87.50 ( 87.22)	Acc@5 100.00 ( 98.86)
Test occlusion: [20/91]	Time  0.252 ( 0.278)	Loss 9.0748e-01 (3.7862e-01)	Acc@1  81.25 ( 89.14)	Acc@5  96.88 ( 99.11)
Test occlusion: [30/91]	Time  0.253 ( 0.270)	Loss 5.9461e-01 (5.3701e-01)	Acc@1  84.38 ( 84.63)	Acc@5 100.00 ( 98.54)
Test occlusion: [40/91]	Time  0.253 ( 0.266)	Loss 3.6392e-01 (5.0611e-01)	Acc@1  92.19 ( 85.33)	Acc@5 100.00 ( 98.82)
Test occlusion: [50/91]	Time  0.252 ( 0.264)	Loss 4.2008e-02 (4.1815e-01)	Acc@1  96.88 ( 87.84)	Acc@5 100.00 ( 99.05)
Test occlusion: [60/91]	Time  0.253 ( 0.262)	Loss 1.5068e-01 (3.7750e-01)	Acc@1  98.44 ( 88.93)	Acc@5 100.00 ( 99.21)
Test occlusion: [70/91]	Time  0.252 ( 0.261)	Loss 6.0342e-01 (5.2264e-01)	Acc@1  85.94 ( 86.49)	Acc@5 100.00 ( 98.83)
Test occlusion: [80/91]	Time  0.260 ( 0.260)	Loss 1.9888e-01 (5.3517e-01)	Acc@1  89.06 ( 86.11)	Acc@5 100.00 ( 98.80)
Test occlusion: [90/91]	Time  0.082 ( 0.257)	Loss 1.0474e-02 (5.0604e-01)	Acc@1 100.00 ( 86.71)	Acc@5 100.00 ( 98.91)
 *   Acc@1 86.711 Acc@5 98.910
```

The wired thing, when i evaluate model accuracy it's significantly decreased. Is this normal behavior?

```
Current iid performance:  0.08704620462046204
TOP-1@shape: 0.09378185524974515
TOP-1@pose: 0.04837209302325581
TOP-1@texture: 0.21238938053097345
TOP-1@context: 0.12522045855379188
TOP-1@weather: 0.12314540059347182
TOP-1@occlusion: 0.058660667935629004
Mean-TOP-1:  0.11026164264781119
```

**Comment by mmmmimic** [comment]
Dear Khawar, 

You are right, I also had the same experience during the first stage of the competition - the accuracy reported during training does not match the evaluation performance. Since this code was adapted from the official baseline of this competition, I would suggest you run the official classification baseline [here](https://github.com/eccv22-ood-workshop/ROBIN-dataset/commit/d8c27076c1c0a43e7ad7e61c1b2f902067f624be) to see if you can reproduce this problem.

**Comment by khawar-islam** [comment]
Dear Manxi,

I can also reproduce on baseline code. Do you have ant way to solve wired behavior?

**Comment by mmmmimic** [comment]
Dear Khawar, 

I am sorry I haven't had time to look into this problem. You might want to ask the challenge organizer who made this baseline:)

**Comment by khawar-islam** [comment]
Dear Manxi,

I have also sent an email to the organizer but did not get any response. Somehow, I figured out the problem and I just need your favor in your `test.py`. You are generating a single `iid.csv` file and this is correct. How we can generate an individual `.csv` file for `'shape', 'pose', 'texture', 'context', 'weather', 'occlusion'` along with `iid.csv` so that we can compare each nuisance class with its ground truth given?

I have modified `test.py` code and also solved some dependency-related issues:

```
from torch.utils.data import Dataset, DataLoader
import pandas as pd
from torchvision import transforms
from PIL import Image
import torch
import torch.nn as nn
from glob import glob
from pathlib import PurePath
import numpy as np
import timm
import torchvision
import time

img_list = glob('/media/cvpr/CM_22/ROBINv1.1-cls-pose/iid_test/Images/*.jpg')

name_list = [
    'aeroplane',
    'bicycle',
    'boat',
    'bus',
    'car',
    'chair',
    'diningtable',
    'motorbike',
    'sofa',
    'train'
]


# conda install pytorch==1.12.1 torchvision==0.13.1 torchaudio==0.12.1 -c pytorch

class PoseData(Dataset):
    def __init__(self, transforms) -> None:
        """
        the data folder should look like
        - datafolder
            - Images
            - labels.csv        
        """
        super().__init__()
        self.img_list = glob('/media/cvpr/CM_22/ROBINv1.1-cls-pose/iid_test/Images/*.jpg')
        self.img_list = sorted(self.img_list, key=lambda x: PurePath(x).parts[-1][:-4])
        self.trs = transforms

    def __len__(self):
        return len(self.img_list)

    def __getitem__(self, index):
        image_dir = self.img_list[index]
        image_name = PurePath(image_dir).parts[-1]
        image = Image.open(image_dir)
        image = self.trs(image)

        return image, image_name


if __name__ == "__main__":
    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     std=[0.229, 0.224, 0.225])
    tfs = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        normalize,
    ])

    model = timm.models.resnet152(pretrained=True, num_classes=10)
    #model = torch.nn.DataParallel(model)
    model.load_state_dict(torch.load('resnet152_best.pth.tar')['state_dict'], strict=False)
    model = model.cuda()
    model.eval()

    dataset = PoseData(tfs)
    loader = DataLoader(dataset, batch_size=128, shuffle=False, drop_last=False, num_workers=4)

    image_dir = []
    preds = []
    for image, pth in loader:
        image_dir.append(list(pth))
        image = image.cuda()

        with torch.no_grad():

            model.eval()
            pred = model(image)

            #pred = torch.softmax(pred[:, :10], dim=1)
            pred = torch.argmax(pred[:, :10], dim=1)

            p = []
            for i in range(pred.size(0)):
                p.append(name_list[pred[i].item()])
        p = np.array(p)
        preds.append(p)
        print(len(np.concatenate(preds)))

    image_dir = np.array(sum(image_dir, []))
    preds = np.concatenate(preds)

    csv = {'imgs': np.array(image_dir), 'pred': np.array(preds),
           }
    csv = pd.DataFrame(csv)
    print(csv)

    csv.to_csv('evaluation/cls_ref/res/iid.csv', index=False)
```

**Comment by mmmmimic** [comment]
Dear Khawar, Yes, you are right. The prediction given by test.py is correct, and we used that for the competition.  You may modify the “iid_test” here to fit the code to “shape”, “pose”, etc.  Have a nice December!  Cheers,Manxi From: Khawar IslamSent: Thursday, December 1, 2022 2:20 AMTo: mmmmimic/OOD-CV-challengeCc: Manxi Lin; CommentSubject: Re: [mmmmimic/OOD-CV-challenge] Wired results on nuisances classes (Issue #2) Dear Manxi,I have also sent an email to the organizer but did not get any response. Somehow, I figured out the problem and I just need your favor in your test.py. You are generating a single iid.csv file and this is correct. How we can generate an individual .csv file for 'shape', 'pose', 'texture', 'context', 'weather', 'occlusion' along with iid.csv so that we can compare each nuisance class with its ground truth given?I have modified test.py code and also solved some dependency-related issues:from torch.utils.data import Dataset, DataLoaderimport pandas as pdfrom torchvision import transformsfrom PIL import Imageimport torchimport torch.nn as nnfrom glob import globfrom pathlib import PurePathimport numpy as npimport timmimport torchvisionimport time img_list = glob('/media/cvpr/CM_22/ROBINv1.1-cls-pose/iid_test/Images/*.jpg') name_list = [    'aeroplane',    'bicycle',    'boat',    'bus',    'car',    'chair',    'diningtable',    'motorbike',    'sofa',    'train']  # conda install pytorch==1.12.1 torchvision==0.13.1 torchaudio==0.12.1 -c pytorch class PoseData(Dataset):    def __init__(self, transforms) -> None:        """        the data folder should look like        - datafolder            - Images            - labels.csv                """        super().__init__()        self.img_list = glob('/media/cvpr/CM_22/ROBINv1.1-cls-pose/iid_test/Images/*.jpg')        self.img_list = sorted(self.img_list, key=lambda x: PurePath(x).parts[-1][:-4])        self.trs = transforms     def __len__(self):        return len(self.img_list)     def __getitem__(self, index):        image_dir = self.img_list[index]        image_name = PurePath(image_dir).parts[-1]        image = Image.open(image_dir)        image = self.trs(image)         return image, image_name  if __name__ == "__main__":    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],                                     std=[0.229, 0.224, 0.225])    tfs = transforms.Compose([        transforms.Resize((224, 224)),        transforms.ToTensor(),        normalize,    ])     model = timm.models.resnet152(pretrained=True, num_classes=10)    #model = torch.nn.DataParallel(model)    model.load_state_dict(torch.load('resnet152_best.pth.tar')['state_dict'], strict=False)    model = model.cuda()    model.eval()     dataset = PoseData(tfs)    loader = DataLoader(dataset, batch_size=128, shuffle=False, drop_last=False, num_workers=4)     image_dir = []    preds = []    for image, pth in loader:        image_dir.append(list(pth))        image = image.cuda()         with torch.no_grad():             model.eval()            pred = model(image)             #pred = torch.softmax(pred[:, :10], dim=1)            pred = torch.argmax(pred[:, :10], dim=1)             p = []            for i in range(pred.size(0)):                p.append(name_list[pred[i].item()])        p = np.array(p)        preds.append(p)        print(len(np.concatenate(preds)))     image_dir = np.array(sum(image_dir, []))    preds = np.concatenate(preds)     csv = {'imgs': np.array(image_dir), 'pred': np.array(preds),           }    csv = pd.DataFrame(csv)    print(csv)     csv.to_csv('evaluation/cls_ref/res/iid.csv', index=False)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>
